### PR TITLE
regression test for #368

### LINKED ISSUE #368: OSError: raw readinto() returned invalid length -1 (should have been between 0 and 8192)
After running service for a while, this error occurred. Not sure if you can reproduce without incoming requests, but it happens to me every time.

Here's my code

``` python
class FetchService:
    name = "fetch_userdata_service"

    @timer(interval=86400)
    def update_token(self):
        # update token every day
        if os.path.isfile(TOKEN_FILE):
            os.remove(os.path.join(current_path, TOKEN_FILE))
            print("Old token removed")
        client.login_in_terminal(username, password)
        client.save_token('token.pkl')
        print("New token saved")

    @http('GET', '/ctrl/stop')
    def stop_service(self, request):
        runner.kill()

    @http('GET', '/<string:userhash>')
    def get_method(self, request, userhash):
        print("request userhash: " + userhash)
        u = client.people(userhash)
        # some lines ignored
        return json.dumps(data, separators=(',', ':'), ensure_ascii=False)

if __name__ == '__main__':
    logger.info("\n\nNameko Service Start...\n")
    runner = ServiceRunner(config={})
    runner.add_service(FetchService)
    runner.start()
    runner.wait()
```

The `client` is something similar to an HTTP client that make requests. And here's error log:

``` bash
INFO:__main__:

Nameko Service Start...

INFO:nameko.runners:starting services: fetch_userdata_service
ERROR:nameko.containers:<ServiceContainer [fetch_userdata_service] at 0x7fce53b92ac8> thread exited with error
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/nameko/containers.py", line 477, in _handle_thread_exited
    gt.wait()
  File "/usr/local/lib/python3.5/dist-packages/eventlet/greenthread.py", line 175, in wait
    return self._exit_event.wait()
  File "/usr/local/lib/python3.5/dist-packages/eventlet/event.py", line 125, in wait
    current.throw(*self._exc)
  File "/usr/local/lib/python3.5/dist-packages/eventlet/greenthread.py", line 214, in main
    result = function(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/eventlet/wsgi.py", line 719, in process_request
    proto.__init__(sock, address, self)
  File "/usr/lib/python3.5/socketserver.py", line 681, in __init__
    self.handle()
  File "/usr/lib/python3.5/http/server.py", line 424, in handle
    self.handle_one_request()
  File "/usr/local/lib/python3.5/dist-packages/eventlet/wsgi.py", line 330, in handle_one_request
    self.raw_requestline = self.rfile.readline(self.server.url_length_limit)
OSError: raw readinto() returned invalid length -1 (should have been between 0 and 8192)
INFO:nameko.containers:killing <ServiceContainer [fetch_userdata_service] at 0x7fce53b92ac8> due to raw readinto() returned invalid length -1 (should have been between 0 and 8192)
WARNING:nameko.containers:killing 2 managed thread(s)
WARNING:nameko.containers:killing managed thread `run`
WARNING:nameko.containers:killing managed thread `<unknown>`
INFO:nameko.runners:stopping services: fetch_userdata_service
CRITICAL:__main__:Uncaught exception
Traceback (most recent call last):
  File "service.py", line 89, in <module>
    runner.wait()
  File "/usr/local/lib/python3.5/dist-packages/nameko/runners.py", line 115, in wait
    SpawningProxy(self.containers, abort_on_error=True).wait()
  File "/usr/local/lib/python3.5/dist-packages/nameko/utils.py", line 179, in spawning_method
    return list(fail_fast_imap(pool, call, self._items))
  File "/usr/local/lib/python3.5/dist-packages/nameko/utils.py", line 148, in fail_fast_imap
    eventlet.getcurrent().throw(*exc_info)
  File "/usr/local/lib/python3.5/dist-packages/nameko/utils.py", line 127, in handle_result
    thread_result = finished_thread.wait()
  File "/usr/local/lib/python3.5/dist-packages/eventlet/greenthread.py", line 175, in wait
    return self._exit_event.wait()
  File "/usr/local/lib/python3.5/dist-packages/eventlet/event.py", line 125, in wait
    current.throw(*self._exc)
  File "/usr/local/lib/python3.5/dist-packages/eventlet/greenthread.py", line 214, in main
    result = function(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/nameko/utils.py", line 176, in call
    return getattr(item, name)(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/nameko/containers.py", line 321, in wait
    return self._died.wait()
  File "/usr/local/lib/python3.5/dist-packages/eventlet/event.py", line 121, in wait
    return hubs.get_hub().switch()
  File "/usr/local/lib/python3.5/dist-packages/eventlet/hubs/hub.py", line 294, in switch
    return self.greenlet.switch()
  File "/usr/local/lib/python3.5/dist-packages/nameko/containers.py", line 477, in _handle_thread_exited
    gt.wait()
  File "/usr/local/lib/python3.5/dist-packages/eventlet/greenthread.py", line 175, in wait
    return self._exit_event.wait()
  File "/usr/local/lib/python3.5/dist-packages/eventlet/event.py", line 125, in wait
    current.throw(*self._exc)
  File "/usr/local/lib/python3.5/dist-packages/eventlet/greenthread.py", line 214, in main
    result = function(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/eventlet/wsgi.py", line 719, in process_request
    proto.__init__(sock, address, self)
  File "/usr/lib/python3.5/socketserver.py", line 681, in __init__
    self.handle()
  File "/usr/lib/python3.5/http/server.py", line 424, in handle
    self.handle_one_request()
  File "/usr/local/lib/python3.5/dist-packages/eventlet/wsgi.py", line 330, in handle_one_request
    self.raw_requestline = self.rfile.readline(self.server.url_length_limit)
OSError: raw readinto() returned invalid length -1 (should have been between 0 and 8192)
```

So you might think there's some bug in my code, but the service would work normally for hours, receiving requests and returning data without any problem. And suddenly it just crashed.

### PR DESCRIPTION
**Pull request by mattbennett**
Fixes #368 

@davidszotten the fix required a change to a test which seemed to explicitly expect a `socket.error` to take down the container (called `test_other_error`). What was the intent of that test?

### CODE REVIEW

File: test/web/test_server.py [hunk 1]
```diff
@@ -43,6 +40,28 @@ def test_broken_pipe(
...
+    """ Regression for https://github.com/onefinestay/nameko/issues/368
+    """
+    container = container_factory(ExampleService, web_config)
+    container.start()
+
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s.connect(('127.0.0.1', web_config_port))
+
+    with patch.object(HttpOnlyProtocol, 'handle_one_request') as handle:
+        handle.side_effect = OSError('raw readinto() returned invalid length')
```
Comment (mattbennett) [inline test/web/test_server.py]: It's difficult to demonstrate that something *doesn't* crash when the crash cause is non-deterministic.

I have something that seems to work though. Do you think the following is preferable?

``` python
import eventlet
import pytest
import requests
from eventlet.wsgi import Server

from nameko.testing.waiting import wait_for_call
from nameko.web.handlers import http


@pytest.fixture
def container(container_factory):

    class Service(object):
        name = "service"

        @http('GET', '/answer')
        def answer(self, request):
            return "42"

    container = container_factory(Service, {})
    container.start()

    return container


def test_foo(container):

    run = [True]

    def wait_for_os_error(args, kwargs, res, exc_info):
        if exc_info and exc_info[0] == OSError:
            del run[:]
            return True

    with eventlet.Timeout(1):
        with wait_for_call(Server, 'process_request', callback=wait_for_os_error):
            while run:
                gt = eventlet.spawn(requests.get, "http://localhost:8000/answer")
                eventlet.sleep(0.01)
                gt.kill()

    # still works
    assert requests.get("http://localhost:8000/answer").text == "42"

```

Comment (mattbennett) [inline test/web/test_server.py]: You might need to show me...

**Comment by mattbennett** [comment]
For a regression test that we don't expect to live very long, I'm not sure it's that helpful

**Comment by davidszotten** [comment]
the eventlet release cadence is very jerky. and even if they release a version with this fix, are we going to set that as the minimum required version for nameko?

**Comment by mattbennett** [comment]
Good point. We wouldn't want to do that until the release has proved itself to be stable.

Do you have something in mind?

**Comment by davidszotten** [comment]
not yet. but given the bug, i think what we need is a socket with `closed` (or `is_closed`, don't recall now) `False`, but that raises an error with one of the error codes considered `SOCKET_CLOSED` https://github.com/davidszotten/eventlet/blob/_recv_loop_socket_closed/eventlet/greenio/base.py#L71

File: test/web/test_server.py (hunk 1 above)
Comment (davidszotten) [inline test/web/test_server.py]: actually, now that we understand this, and know this is (c)python complaining about a protocol violation by eventlet, i'm happy to just simulate that exact response

File: nameko/web/server.py [hunk 2]
```diff
@@ -76,9 +76,21 @@ def run(self):
...
+    def process_request(self, sock, address):
+        try:
+            self._serv.process_request((sock, address))
+        except OSError as exc:
+            # OSError("raw readinto() returned invalid length")
+            # can be raised when a client disconnects very early as a result
+            # of an eventlet bug: https://github.com/eventlet/eventlet/pull/353
+            # See https://github.com/onefinestay/nameko/issues/368
+            if "raw readinto() returned invalid length" in str(exc):
+                return
```
Comment (davidszotten) [inline nameko/web/server.py]: no cleanup needed here?

Comment (mattbennett) [inline nameko/web/server.py]: I don't think so. The managed thread will exit without error and the server will continue waiting for another `accept()`

**Pull request merged**